packages/next/src/build/webpack/plugins/build-manifest-plugin.ts

```diff
@@ -273,15 +273,88 @@ export default class BuildManifestPlugin {
         if (SYSTEM_ENTRYPOINTS.has(entrypoint.name)) continue
         const pagePath = getRouteFromEntrypoint(entrypoint.name)
 
-        if (!pagePath) {
-          continue
-        }
+        if (!pagePath) continue
 
         const filesForPage = getEntrypointFiles(entrypoint)
-
         assetMap.pages[pagePath] = [...new Set([...mainFiles, ...filesForPage])]
       }
 
+      // When a CSS file is server-rendered, it is cleaned up during
+      // the client render. During navigation, if the target route
+      // dynamically imports the same CSS file, the client will skip
+      // loading it as it was already loaded statically, but still
+      // cleaned up during the client render. This results in missing
+      // the expected CSS styles.
+      // x-ref: `next/src/client/index.tsx`
+
+      // 1. page transition
+      // 2. import (CSS already exists)
+      // 3. client render starts
+      // 4. clean up server-rendered CSS
+      // 5. CSS is missing
+
+      // To prevent this, we add the dynamically imported CSS files
+      // to the build manifest to signal the client not to clean up
+      // the CSS file if it's loaded dynamically, preserving the styles.
+
+      /* Inspired by `next/src/build/webpack/react-loadable-plugin.ts` */
+      const _handleBlock = (block: any) => {
+        block.blocks.forEach(_handleBlock)
+
+        const dynamicCSSFiles = new Set<string>()
+
+        const chunkGroup = compilation.chunkGraph.getBlockChunkGroup(block)
+        if (chunkGroup) {
+          for (const chunk of chunkGroup.chunks) {
+            chunk.files.forEach((file: string) => {
+              if (file.endsWith('.css') && file.match(/^static\/css\//)) {
+                dynamicCSSFiles.add(file)
+              }
+            })
+          }
+        }
+
+        for (const dependency of block.dependencies) {
+          // We target the two types of dynamic import() dependencies:
+          // `ImportDependency` and `ContextElementDependency`.
+
+          // ImportDependency:
+          // - syntax: import("./module")
+          // - dependency.type: "import()"
+
+          // ContextElementDependency:
+          // - syntax: import(`./module/${param}`)
+          // - dependency.type: "import() context element"
+          if (dependency.type.startsWith('import()')) {
+            const parentModule =
+              compilation.moduleGraph.getParentModule(dependency)
+
+            const parentChunks =
+              compilation.chunkGraph.getModuleChunks(parentModule)
+
+            for (const chunk of parentChunks) {
+              const chunkName = chunk.name
+              if (!chunkName || !chunkName.startsWith('pages')) continue
+
+              const pagePath = getRouteFromEntrypoint(chunkName)
+              if (!pagePath) continue
+
+              const assetMapPages = assetMap.pages[pagePath]
+              // We should already have `assetMapPages` from the `entrypoint`.
```

Comment:
Instead of only saying "we should xxx" maybe also give some explanation why is that.

Reply:
Changed to:

```
// As we run this function `addDynamicCSSToBuildManifest`
// after the entrypoint files are added to the manifest,
// `assetMapPages` should already exist.
```

`entrypoint files` meaning here: https://github.com/vercel/next.js/pull/68396/files#diff-3799047246d3b72b0ae57419b0e4e0a3e192ee8e35cf924d81ff860749fd65b5R340-R348